### PR TITLE
[Models][QwenVL] Remove unnecessary `.contiguous()` calls

### PR DESCRIPTION
**Pull request by lgeiger**
## Purpose

This PR removes calls to `.contiguous()` in the `QwenVisionAttention` module. Tested with both xformers and flash attention backend.

## Test Plan

```
vllm bench serve Qwen/Qwen2.5-VL-3B-Instruct

vllm bench serve --backend openai-chat --model Qwen/Qwen2.5-VL-3B-Instruct --endpoint /v1/chat/completions --dataset-name hf --dataset-path lmarena-ai/VisionArena-Chat --hf-split train --num-prompts 1000
```

### Before

Screenshot 2025-10-17 at 16 05 07
Screenshot 2025-10-17 at 16 06 26

```
============ Serving Benchmark Result ============
Successful requests:                     998
Failed requests:                         2
Benchmark duration (s):                  100.72
Total input tokens:                      94309
Total generated tokens:                  105944
Request throughput (req/s):              9.91
Output token throughput (tok/s):         1051.90
Peak output token throughput (tok/s):    5104.00
Peak concurrent requests:                998.00
Total Token throughput (tok/s):          1988.28
---------------Time to First Token----------------
Mean TTFT (ms):                          44047.03
Median TTFT (ms):                        43652.87
P99 TTFT (ms):                           95956.25
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          199.36
Median TPOT (ms):                        217.51
P99 TPOT (ms):                           252.47
---------------Inter-token Latency----------------
Mean ITL (ms):                           199.33
Median ITL (ms):                         220.68
P99 ITL (ms):                            437.93
==================================================
```

### After

```
============ Serving Benchmark Result ============
Successful requests:                     998
Failed requests:                         2
Benchmark duration (s):                  100.23
Total input tokens:                      94290
Total generated tokens:                  106295
Request throughput (req/s):              9.96
Output token throughput (tok/s):         1060.55
Peak output token throughput (tok/s):    4899.00
Peak concurrent requests:                998.00
Total Token throughput (tok/s):          2001.33
---------------Time to First Token----------------
Mean TTFT (ms):                          43825.06
Median TTFT (ms):                        42747.21
P99 TTFT (ms):                           95296.01
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          197.58
Median TPOT (ms):                        213.70
P99 TPOT (ms):                           252.91
---------------Inter-token Latency----------------
Mean ITL (ms):                           197.16
Median ITL (ms):                         218.65
P99 ITL (ms):                            462.42
==================================================
```

### CODE REVIEW

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 1]
```diff
@@ -396,7 +396,7 @@ def forward(
         q, k, v = self.split_qkv(x)
         batch_size = q.shape[1]
 
-        q, k, v = (rearrange(x, "s b ... -> b s ...").contiguous() for x in (q, k, v))
+        q, k, v = (rearrange(x, "s b ... -> b s ...") for x in (q, k, v))
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply vllm/model_executor/models/qwen2_5_vl.py:399]: I verified that this works fine with the upstream flash attention backend as well:
```
============ Serving Benchmark Result ============
Successful requests:                     998
Failed requests:                         2
Benchmark duration (s):                  99.13
Total input tokens:                      94320
Total generated tokens:                  105913
Request throughput (req/s):              10.07
Output token throughput (tok/s):         1068.38
Peak output token throughput (tok/s):    4774.00
Peak concurrent requests:                998.00
Total Token throughput (tok/s):          2019.82
---------------Time to First Token----------------
Mean TTFT (ms):                          43090.79
Median TTFT (ms):                        42479.09
P99 TTFT (ms):                           94087.45
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          195.36
Median TPOT (ms):                        211.59
P99 TPOT (ms):                           250.32
---------------Inter-token Latency----------------
Mean ITL (ms):                           195.34
Median ITL (ms):                         218.20
P99 ITL (ms):                            450.43
==================================================
```

**Comment by tjtanaa** [comment]
Evaluated on ROCm

Server command: `vllm serve --model Qwen/Qwen2.5-VL-3B-Instruct`


Evaluation score on chartqa
```

For detailed information on this command, run:
  run.py eval_vllm --model_name Qwen/Qwen2.5-VL-3B-Instruct --url http://0.0.0.0:8000 --output_dir ./chartqa --eval_name chartqa - --help
================================================================================
Metrics:
{
    "explicit_prompt_relaxed_correctness": 0.8108,
    "anywhere_in_answer_relaxed_correctness": 0.8144
}
================================================================================
```

**Review by DarkLight1337 — APPROVED** [review]
Thanks for improving this

**Comment by JartX** [comment]
@lgeiger Hi! With this PR Qwen/Qwen3-VL-30B-A3B-Instruct, an invoice is considered to be a cheesecake or a sushi restaurant.
https://github.com/vllm-project/llm-compressor/issues/1939#issuecomment-3425386080

**Comment by JartX** [comment]
@lgeiger Thank you for your quick response, it this happens when I send the image via request, for example via openwebui.

**Comment by JartX** [comment]
@lgeiger my docker run command

docker run -it --rm \
  --name vllm \
  --tty \
  --restart unless-stopped \
  --shm-size=48gb \
  -p 80:8000 \
  -v $(pwd)/app:/app \
  -v $(pwd)/models:/models \
  -v $(pwd)/huggecache:/root/.cache/huggingface \
  --device /dev/kfd \
  --device /dev/dri \
  --group-add video \
  --ipc=host \
  --network host \
  --cap-add SYS_PTRACE \
  --security-opt seccomp=unconfined \
  --privileged \
  -e HSA_OVERRIDE_GFX_VERSION=11.0.0 \
  -e VLLM_USE_V1=1 \
  -e VLLM_V1_USE_PREFILL_DECODE_ATTENTION=1 \
  vllm-rocm-251021-revert \
  vllm serve Qwen/Qwen3-VL-30B-A3B-Instruct \
  --gpu-memory-utilization 0.97 \
  --max_model_len 40960 \
  -tp 4 \
  --served-model-name QWEN3VL \
  --port 80 \
  --enable-auto-tool-choice \
  --compilation-config '{"full_cuda_graph": true}' \
  --disable-log-requests \
  --tool-call-parser hermes \
  --dtype float16 \

ROCM RDNA3

**Comment by lgeiger** [comment]
@JartX To double check, can you reproduce the lm eval results that I posted above or do you also see a regression with and without this PR?

**Comment by JartX** [comment]
Using..
 VLLM_WORKER_MULTIPROC_METHOD=spawn VLLM_ATTENTION_BACKEND=TORCH_SDPA lm_eval --model vllm-vlm --model_args "pretrained=Qwen/Qwen3-VL-30B-A3B-Instruct,max_model_len=10000,tensor_parallel_size=4" --tasks chartqa --batch_size auto --apply_chat_template
Processed prompts:   2%| | 53/2500 [16:43<9:05:51, 13.38s/it, est. speed input: 36.35 toks/s, output: 9
I go to try launch first the vllm serve and after lm_eval api

**Comment by tjtanaa** [comment]
@lgeiger This issue will need to be debugged on AMD GPUs.

I remember @DarkLight1337 mentioned that on ROCm we might need the `.contiguous()`. I last verified on mi300x without this `.contiguous()` and does not seem to have problem. Maybe it is because I was testing with _Backend.FLASH_ATTN and _Backend.ROCM_AITER_FA. Could be that if we use _Backend.TORCH_SDPA, we need to cast `.contiguous()`.

However, I am afraid I might have missed some edge cases that @JartX found.

**Comment by JartX** [comment]
Without Contiguous
local-chat-completions (model=QWEN3VL,base_url=http://localhost:8000/v1/chat/completions,num_concurrent=10,max_retries=3,tokenized_requests=False), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: auto
| Tasks |Version|Filter|n-shot|     Metric      |   |Value |   |Stderr|
|-------|------:|------|-----:|-----------------|---|-----:|---|-----:|
|chartqa|      0|none  |     0|anywhere_accuracy|↑  |0.8712|±  |0.0067|
|       |       |none  |     0|exact_match      |↑  |0.6380|±  |0.0096|
|       |       |none  |     0|relaxed_accuracy |↑  |0.8604|±  |0.0069|

With Contiguos
local-chat-completions (model=QWEN3VL,base_url=http://localhost:8000/v1/chat/completions,num_concurrent=10,max_retries=3,tokenized_requests=False), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: auto
| Tasks |Version|Filter|n-shot|     Metric      |   |Value |   |Stderr|
|-------|------:|------|-----:|-----------------|---|-----:|---|-----:|
|chartqa|      0|none  |     0|anywhere_accuracy|↑  |0.0408|±  |0.0040|
|       |       |none  |     0|exact_match      |↑  |0.0272|±  |0.0033|
|       |       |none  |     0|relaxed_accuracy |↑  |0.0404|±  |0.0039|

@lgeiger @tjtanaa 
I would say that it varies quite a bit

**Comment by tjtanaa** [comment]
@JartX which ViT attention backend did you test? TORCH_SDPA?

**Comment by JartX** [comment]
@lgeiger @tjtanaa Hey guys, I see there are more reports of this problem. Could we try to fix it? Thanks a lot :)

**Comment by lgeiger** [comment]
> @lgeiger @tjtanaa Hey guys, I see there are more reports of this problem. Could we try to fix it? Thanks a lot :)

I don't have access to an AMD GPU, so would be great if either @JartX or @tjtanaa could make a PR with a fix.

**Comment by JartX** [comment]
> > @lgeiger @tjtanaa Hey guys, I see there are more reports of this problem. Could we try to fix it? Thanks a lot :)
> 
> I don't have access to an AMD GPU, so would be great if either @JartX or @tjtanaa could make a PR with a fix.

@lgeiger @tjtanaa 
If it's absolutely necessary to remove the contiguous, we could filter by the attention backend, to use it or not. Another option would be to do a revert. I'd bet, if you've tried and it improves performance, the solution I just mentioned. Or simply add a conditional: if it's in on_gfx9, use the contiguous, because it will be using an attention backend higher than SDPA.

**Comment by tjtanaa** [comment]
@DarkLight1337 @JartX 

I have validated on AMD Instinct GPU. So generally on ROCm if we are using TORCH_SDPA backend we need to cast to contiguous().
However, we shouldn't revert the PR as it is only need when both conditions `is on rocm` and `TORCH_SDPA`.

```diff
        elif self.attn_backend == _Backend.TORCH_SDPA:
            # Execute attention entry by entry for speed & less VRAM.
            from vllm.platforms import current_platform
+            if current_platform.is_rocm():
+                q = q.contiguous()
+                k = k.contiguous()
+                v = v.contiguous()
            outputs = []
            for i in range(1, len(cu_seqlens)):
                start_idx = cu_seqlens[i - 1]
                end_idx = cu_seqlens[i]
                q_i = q[:, start_idx:end_idx]
                k_i = k[:, start_idx:end_idx]
                v_i = v[:, start_idx:end_idx]
                q_i, k_i, v_i = (
                    rearrange(x, "b s h d -> b h s d") for x in [q_i, k_i, v_i]
                )
                output_i = F.scaled_dot_product_attention(q_i, k_i, v_i, dropout_p=0.0)
                output_i = rearrange(output_i, "b h s d -> b s h d ")
                outputs.append(output_i)
            context_layer = torch.cat(outputs, dim=1)
            context_layer = rearrange(
                context_layer, "b s h d -> s b (h d)"
            ).contiguous()
```

So instead of placing at the usual place, we should only cast to contiguous in the `TORCH SDPA` path.

However, at current moment, ROCm will never use TORCH_SDPA. We also need this PR https://github.com/vllm-project/vllm/pull/27190?notification_referrer_id=NT_kwDOAb0gkLQxOTc3Mjk3ODUxMzoyOTE3MTg1Ng .

The current optimization is only done for `qwen2_5_vl.py`. However, I believe this `Remove unnecessary .contiguous() calls` could be applied to many other `models.py` file:
- `qwen2_5_vl.py` (Done by this PR)
- `qwen2_vl.py`
- `ernie45_vl.py`
- `glm4_1v.py`
- `keye.py`

**Comment by JartX** [comment]
@tjtanaa  So I was wrong, sorry, I thought so hahaha, that way we'd both appear at once. Also, do you mind if I put you as a coauthor of the commit?

**Comment by tjtanaa** [comment]
@JartX Sure. Thank you so much for considering me as the coauthor of the commit. 
I am mainly contributing to the AMD features and bugfix on Instinct GPU and some vision model related features.

**Comment by JartX** [comment]
@tjtanaa Of course, all together! 💪
